Brokerages/Oanda/OandaBrokerage.DataQueueHandler.cs

```diff
@@ -84,27 +94,74 @@ public void Subscribe(LiveNodePacket job, IEnumerable<Symbol> symbols)
-                return;
+            lock (_lockerSubscriptions)
+            {
+                var symbolsToUnsubscribe = (from symbol in symbols
+                                            where _subscribedSymbols.Contains(symbol)
+                                            select symbol).ToList();
+                if (symbolsToUnsubscribe.Count == 0)
+                    return;
+
+                Log.Trace("OandaBrokerage.Unsubscribe(): {0}", string.Join(",", symbolsToUnsubscribe.Select(x => x.Value)));
+
+                // Oanda does not allow more than a few rate streaming sessions, 
+                // so we only use a single session for all currently subscribed symbols
+                var symbolsToSubscribe = _subscribedSymbols.ToList().Where(x => !symbolsToUnsubscribe.Contains(x)).ToList();
 
-            Log.Trace("OandaBrokerage.Unsubscribe(): {0}", string.Join(",", symbolsToUnsubscribe.Select(x => x.Value)));
+                _subscribedSymbols = symbolsToSubscribe.ToHashSet();
 
-            // Oanda does not allow more than a few rate streaming sessions, 
-            // so we only use a single session for all currently subscribed symbols
-            var symbolsToSubscribe = _subscribedSymbols.ToList().Where(x => !symbolsToUnsubscribe.Contains(x)).ToList();
+                ProcessSubscriptionRequest();
+            }
+        }
 
-            SubscribeSymbols(symbolsToSubscribe);
+        /// <summary>
+        /// Groups multiple subscribe/unsubscribe calls to avoid closing and reopening the streaming session on each call
+        /// </summary>
+        private void ProcessSubscriptionRequest()
+        {
+            if (_subscriptionsPending) return;
+            
+            _lastSubscribeRequestUtcTime = DateTime.UtcNow;
+            _subscriptionsPending = true;
 
-            _subscribedSymbols = symbolsToSubscribe.ToHashSet();
+            Task.Run(() =>
+            {
+                while (true)
+                {
+                    DateTime requestTime;
+                    List<Symbol> symbolsToSubscribe;
+                    lock (_lockerSubscriptions)
+                    {
+                        requestTime = _lastSubscribeRequestUtcTime.Add(SubscribeDelay);
+                        symbolsToSubscribe = _subscribedSymbols.ToList();
+                    }
+
+                    if (DateTime.UtcNow > requestTime)
+                    {
+                        // restart streaming session
+                        SubscribeSymbols(symbolsToSubscribe);
+
+                        lock (_lockerSubscriptions)
+                        {
+                            _lastSubscribeRequestUtcTime = DateTime.UtcNow;
+                            if (_subscribedSymbols.Count == symbolsToSubscribe.Count)
+                            {
+                                // no more subscriptions pending, task finished
+                                _subscriptionsPending = false;
```

Comment:
Merged in https://github.com/QuantConnect/Lean/commit/3b8e0c64f0fb9be3c5eac2cabccec7ffc020296d